In [ ]:
from sklearn.datasets import fetch_california_housing

# Load the California housing dataset
housing = fetch_california_housing()

In [4]:
# Create a Pandas DataFrame for the features and a Series for the target variable (med_house_value)
import pandas as pd
X = pd.DataFrame(housing.data, columns=housing.feature_names)
y = pd.Series(housing.target, name='med_house_value')

In [8]:
# Perform an initial exploration of the dataset:
# Display the first five rows of the dataset
print(X.head())
# Print the feature names and check for missing values
print("Feature Names:")
print(X.columns.tolist())
print("\nMissing Values:")
print(X.isnull().sum())
# Generate summary statistics (mean, min, max, etc.)
print("\nSummary Statistics:")
print(X.describe())

   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  
Feature Names:
['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']

Missing Values:
MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

Summary Statistics:
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
me

# Part 2

In [12]:
# Split the dataset into training and test sets (80% training, 20% test)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Train a linear regression model on the unscaled data using sklearn.linear_model.LinearRegression
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)
# Make predictions on the test set
y_pred = model.predict(X_test)
# Evaluate the model's performance using the following metrics: Mean Squared Error, Root Mean Squared Error, and R-squared score
from sklearn.metrics import mean_squared_error, r2_score
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"Root Mean Squared Error: {rmse}")
print(f"R-squared Score: {r2}")

# What are the coefficients of the linear regression model
coefficients = pd.Series(model.coef_, index=X.columns)
print("\nCoefficients of the Linear Regression Model:")
print(coefficients)

Mean Squared Error: 0.5558915986952442
Root Mean Squared Error: 0.7455813830127763
R-squared Score: 0.575787706032451

Coefficients of the Linear Regression Model:
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64


Interpretations:
The R-squared score tells us that roughly 58% of the variation in median housing value is explained by our model. While a "good" R-squared value depends on the context, we would like to have a higher score here. 
The variables that seem to have the largest impact on median housing value: Latitude (-.42), Longitude (-.43), AveBedrms (.78), and MedInc (.45). Of all the variables, the average number of bedrooms has the biggest impact on predictions. This makes sense since many nicer houses have a large number of bedrooms/guest bedrooms.
How well do our predicted values match our actual values? We can look at our RMSE to tell that our predictions were, on average, .75 points away from the actual values.


# Part 3

In [15]:
# Select three features from the dataset to build a simplified model.Explain your choice.
# I select: MedInc, AveBedrms, and Longitude. I chose these three because they had the highest coefficients in the initial regression. They also provide a mix of elements that go into housing value (location, owner income, and actual house characteristics).
X_train_simple = X_train[['MedInc', 'AveBedrms', 'Longitude']]
X_test_simple = X_test[['MedInc', 'AveBedrms', 'Longitude']]

# Train a new linear regression model using only the three selected features
model_simple = LinearRegression()
model_simple.fit(X_train_simple, y_train)

# Evaluate the performance of this simplified model and compare it to the full model
y_pred_simple = model_simple.predict(X_test_simple)
mse_simple = mean_squared_error(y_test, y_pred_simple) 
rmse_simple = mse_simple ** 0.5
r2_simple = r2_score(y_test, y_pred_simple)
print("\nPerformance of the Simplified Model:")
print(f"Mean Squared Error: {mse_simple}")
print(f"Root Mean Squared Error: {rmse_simple}")
print(f"R-squared: {r2_simple}")
# Compare this model to the performance of the full model
print("\nComparison to the Full Model:")
print(f"Full Model MSE: {mse}, Simplified Model MSE: {mse_simple}")
print(f"Full Model RMSE: {rmse}, Simplified Model RMSE: {rmse_simple}")
print(f"Full Model R-squared: {r2}, Simplified Model R-squared: {r2_simple}")



Performance of the Simplified Model:
Mean Squared Error: 0.7070831927336602
Root Mean Squared Error: 0.8408823893587379
R-squared: 0.4604102959651172

Comparison to the Full Model:
Full Model MSE: 0.5558915986952442, Simplified Model MSE: 0.7070831927336602
Full Model RMSE: 0.7455813830127763, Simplified Model RMSE: 0.8408823893587379
Full Model R-squared: 0.575787706032451, Simplified Model R-squared: 0.4604102959651172


Interpretations:
The simplified model overall seems to be a poorer fit for our data than the full model. The simple model only accounts for 46% of the variation in median housing value, compared to 58% in the full model. Also, the error outputs are both worse. The average distance between our simple model and the actual values is 0.84 points instead of .75 as in the full model.
I would not use this simple model in practice, unless I was not very concerned about the exactness of the estimates. Leaving our other variables out of the regression, like latitude, average rooms, etc., will leave us with less exact results, and the strength of the coefficients on our simple model variables will likely be skewed. Controlling for more variables helps us have more accurate estimates.